bindings/node/lib/clientEncryption.js

```diff
@@ -549,6 +554,63 @@ module.exports = function (modules) {
       return value;
     }
 
+    /**
+     * Creates a collection that has encrypted document fields
+     *
+     * @template {TSchema} - Schema for the collection being created
+     * @param {Db} db - A database to create the collection in
+     * @param {string} name - The name of the collection to be created
+     * @param {object} options - Options for createDataKey and for createCollection.
+     * @param {string} options.provider - provider name
+     * @param {ClientEncryptionCreateDataKeyProviderOptions} options.createDataKeyOptions - options to pass to createDataKey
+     * @param {CreateCollectionOptions} options.createCollectionOptions - options to pass to createCollection, must include `encryptedFields`
+     * @returns {Promise<{ collection: Collection<TSchema>, encryptedFields: Document }>} - created collection and generated encryptedFields
+     * @throws {MongoCryptCreateEncryptedCollectionError} - If part way through the process createDataKey fails, an error will be rejected that has the `encryptedFields` that were created.
+     */
+    async createEncryptedCollection(db, name, options) {
+      const {
+        provider,
+        createDataKeyOptions = {},
+        createCollectionOptions: {
+          encryptedFields: { ...encryptedFields },
+          ...createCollectionOptions
+        }
+      } = options;
+
+      if (Array.isArray(encryptedFields.fields)) {
+        const createDataKeyPromises = [...encryptedFields.fields].map(async field =>
+          typeof field !== 'object' || field.keyId != null
+            ? field
+            : { ...field, keyId: await this.createDataKey(provider, createDataKeyOptions) }
+        );
+
+        const createDataKeyResolutions = await Promise.allSettled(createDataKeyPromises);
+
+        encryptedFields.fields = createDataKeyResolutions.map((resolution, index) =>
+          resolution.status === 'fulfilled' ? resolution.value : encryptedFields.fields[index]
+        );
+
+        if (createDataKeyResolutions.some(resolution => resolution.status === 'rejected')) {
+          throw new common.MongoCryptCreateEncryptedCollectionError(
+            'Could not complete creating data keys',
+            {
+              encryptedFields,
+              errors: createDataKeyResolutions.map(resolution =>
+                resolution.status === 'rejected' ? resolution.reason : null
+              )
```

Comment:
@baileympearson @addaleax 

This would results in an array with null entries for the operations that did not have an error. Alternatively, we could return:

```ts
errors: createDataKeyResolutions.reduce((acc, resolution, index) =>
  if (resolution.status === 'rejected') acc[index] = resolution.reason;
  return acc;
}, {})
```

An object with indexes as keys that map to the error. 

I think regardless of the decision we make it will bespoke to this function so users will have to rely on the documentation of _how_ to use the errors prop on the `MongoCryptCreateEncryptedCollectionError`

Reply:
And alternatively: 
- returning an array makes all the typical helpers available, map/filter/etc depending what the user needs to do with their error handling
- And the array isn't sparse so it defines all numeric keys btwn `'0'` and length of `encryptedFields.fields`

I'm up for either, let's get one more engineer's input (notably the tests shouldn't need any changes b/c of my `.nested.property` checks, whatever we decide let's remember to assert the expected outcome type)

Reply:
We could also make the array compact if we maintained the index information like so:

```js
const errors = createDataKeyResolutions
  .map((resolution, index) => ({ index, error: resolution.reason }))
  .filter(({error}) => error != null);
```

Reply:
From virtual office @dariakp @baileympearson we discussed taking only the first error. This is updated.